SeQuant/core/eval_expr.cpp

```diff
@@ -1,140 +1,217 @@
+  size_t h = 0;
+  hash::combine(h, hash::value(op));
 
-EvalExpr::EvalExpr(const EvalExpr& xpr1, const EvalExpr& xpr2,
-                   EvalOp op) noexcept
-    : id_{++global_id} {
-  assert(op != EvalOp::Id);
-  auto const& expr1 = xpr1.hash_value() < xpr2.hash_value() ? xpr1 : xpr2;
-  auto const& expr2 = xpr1.hash_value() < xpr2.hash_value() ? xpr2 : xpr1;
-
-  auto bk = braket_type{};
-  if (op == EvalOp::Prod)
-    bk = target_braket_prod(expr1.tensor(), expr2.tensor());
-  else {
-    std::get<0>(bk) = expr1.tensor().bra() | ranges::to<index_container_type>;
-    std::get<1>(bk) = expr1.tensor().ket() | ranges::to<index_container_type>;
-  }
+  auto lh = hash::value(left);
+  auto rh = hash::value(right);
 
-  auto const& t1 = expr1.tensor();
-  auto const& t2 = expr2.tensor();
-  Symmetry s = Symmetry::invalid;
-  switch (op) {
-    case EvalOp::Sum:
-      s = infer_tensor_symmetry_sum(expr1, expr2);
-      break;
-    case EvalOp::Prod:
-      s = infer_tensor_symmetry_prod(expr1, expr2);
-      break;
-    case EvalOp::Symm:
-      s = Symmetry::symm;
-      break;
-    case EvalOp::Antisymm:
-      s = Symmetry::antisymm;
-      break;
-    default:
-      assert(false && "Unsupported operation for symmetry detect.");
-  }
-  op_ = op;
+  hash::combine(h, lh < rh ? lh : rh);
+  hash::combine(h, lh < rh ? rh : lh);
+
+  if (left.result_type() == ResultType::Tensor &&
+      right.result_type() == ResultType::Tensor)
+    hash::combine(h, hash_tensor_pair_topology(left.expr()->as<Tensor>(),
+                                               right.expr()->as<Tensor>()));
+  return h;
+}
+
+std::pair<sequant::container::svector<sequant::Index>,
+          sequant::container::svector<sequant::Index>>
+target_braket(sequant::Tensor const& t1, sequant::Tensor const& t2) noexcept {
+  using ranges::views::keys;
+  using ranges::views::values;
+  using ranges::views::zip;
+  using namespace sequant;
+  using index_container = container::svector<Index>;
+
+  auto remove_item = [](auto& vec, size_t pos) -> void {
+    std::swap(vec[pos], vec.back());
+    vec.pop_back();
+  };
+
+  auto left = zip(t1.bra(), t1.ket()) | ranges::to_vector;
+  auto right = zip(t2.bra(), t2.ket()) | ranges::to_vector;
 
-  tensor_ = Tensor{L"I", std::get<0>(bk),         std::get<1>(bk),
-                   s,    infer_braket_symmetry(), infer_particle_symmetry(s)};
+next_contract:
+  while (!right.empty()) {
+    for (auto rr = 0; rr < right.size(); ++rr) {
+      auto& [rb, rk] = right[rr];
+      for (auto ll = 0; ll < left.size(); ++ll) {
+        auto& [lb, lk] = left[ll];
+        if (lb == rk && rb == lk) {
+          remove_item(left, ll);
+          remove_item(right, rr);
+          goto next_contract;
+        } else if (lb == rk) {
+          std::swap(lk, rk);
+          remove_item(left, ll);
+          goto next_contract;
+        } else if (lk == rb) {
+          std::swap(lb, rb);
+          remove_item(left, ll);
+          goto next_contract;
```

Comment:
To break from a nested loop, you should only jump forward using `goto`, but never backwards: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-goto

Reply:
This is a tricky situation. Also this function will be generalized for contraction involving tensors with unequal number of bra and ket indices. It can/will be addressed then.